File: leetcode/exercises/2.2.3.medium.evaluate_reverse_polish_notation.go
// -----------------------
// Created     :9/11/2025
// Last Edited :9/11/2025 
// Topics      :
// Big O       :
// Problem Id  :50
// References  :
// Notes       :
// -----------------------

package main

import (
    "fmt"
    "strings"
    "strconv"
)

func evalRPN(tokens []string) int {
    stack := []int{}
    res := 0
    for _, v := range tokens {
        if !strings.Contains("+-/*", v) {
            i, err := strconv.Atoi(v)
            if err != nil {
                panic(err)
            }
            stack = append(stack, i)        
            continue
        } else if v == "-" {
            res = stack[len(stack) - 2] - stack[len(stack) - 1]
        } else if v == "*" {
            res = stack[len(stack) - 2] * stack[len(stack) - 1]
        } else if v == "/" {
            res = stack[len(stack) - 2] / stack[len(stack) - 1]
        } else {
            res = stack[len(stack) - 2] + stack[len(stack) - 1]
        }
        stack = stack[:len(stack)-2]
        stack = append(stack, res)
    }
    return stack[0]
}

func main() {
    fmt.Println(evalRPN([]string{"4","13","5","/","+"}))
}

File: leetcode/exercises/2.2.3.medium.evaluate_reverse_polish_notation.py
# -----------------------
# Created     : 19/11/2025
# Last Edited : 19/11/2025 
# Topics      : 
# Big O       :
# Problem Id  : 150
# References  :
# Notes       :
# -----------------------

import math

class Solution:
    def evalRPN(self, tokens: list[str]) -> int:

        stack = []
        for i in tokens:

            if i == "-":
                a = stack.pop()
                b = stack.pop()
                stack.append(b - a)

            elif i == "+":
                stack.append(stack.pop() + stack.pop())

            elif i == "*":
                stack.append(stack.pop() * stack.pop())

            elif i == "/":
                a = stack.pop()
                b = stack.pop()
                stack.append(int(b / a))

            else:
                stack.append(int(i))

        return stack[0]

s = Solution()
print(s.evalRPN(["4","13","5","/","+"]))


                
                


